clang/include/clang/Basic/Attr.td

```diff
@@ -1889,6 +1889,36 @@ def LifetimeBound : DeclOrTypeAttr {
   let SimpleHandler = 1;
 }
 
+def LifetimeCaptureBy : DeclOrTypeAttr {
+  let Spellings = [Clang<"lifetime_capture_by", 0>];
+  let Subjects = SubjectList<[ParmVar, ImplicitObjectParameter], ErrorDiag>;
+  let Args = [VariadicParamOrParamIdxArgument<"Params">];
+  let Documentation = [LifetimeCaptureByDocs];
+  let AdditionalMembers = [{
+private:
+  IdentifierInfo** ArgIdents;
```

Comment:
Not for this PR, but I think one option would be to have a trailing array of pairs, so we don't need multiple allocations but we can embed the arguments in the attribute. Not sure if it is cumbersome to do for attributes, so it is just a random idea for the future.